browser-extension/mv3/src/service-worker/services/tabService.ts

```diff
@@ -0,0 +1,198 @@
+      }
+    });
+
+    chrome.webRequest.onBeforeRequest.addListener(
+      (details) => {
+        if (details.type === "main_frame") {
+          const tab = this.getTab(details.tabId) || ({ id: details.tabId } as TabData);
+          this.addOrUpdateTab({ ...tab, url: details.url });
+        }
+      },
+      { urls: ["<all_urls>"] }
+    );
+
+    chrome.webNavigation.onCommitted.addListener((navigatedTabData) => {
+      const tab = this.getTab(navigatedTabData.tabId);
+
+      if (tab) {
+        this.addOrUpdateTab({
+          ...tab,
+          documentLifecycle:
+            tab.documentLifecycle !== "active" ? navigatedTabData.documentLifecycle : tab.documentLifecycle,
+          frameId: tab.frameId !== 0 ? navigatedTabData.frameId : tab.frameId,
+        });
+      }
+    });
+  }
+
+  private sendMessage(tabId: TabId, ...args: [any, any?, ((response: any) => void)?]) {
+    chrome.tabs.sendMessage(tabId, ...args);
+  }
+
+  addOrUpdateTab(tab: TabData) {
+    // A special ID value given to tabs that are not browser tabs (for example, apps and devtools windows)
+    if (tab.id !== chrome.tabs.TAB_ID_NONE) {
+      this.map[tab.id] = tab;
+    }
+  }
+
+  createNewTab(url: string, openerTabId: TabId, callback: Function) {
+    chrome.tabs.create({ url, openerTabId }, (tab) => {
+      callback(tab);
+    });
+  }
+
+  removeTab(tabId: TabId) {
+    delete this.map[tabId];
+  }
+
+  getTabs() {
+    return this.map;
+  }
+
+  getTab(tabId: TabId) {
+    return this.getTabs()[tabId];
+  }
+
+  getTabUrl(tabId: TabId) {
+    var tab = this.getTab(tabId);
+    return tab && tab.url;
+  }
+
+  focusTab(tabId: TabId) {
+    var tab = this.getTab(tabId);
+
+    if (tab && tab.windowId) {
+      try {
+        chrome.windows.update(tab.windowId, { focused: true }, () => {
+          chrome.tabs.highlight({ windowId: tab.windowId, tabs: tab.index });
+        });
+        return true;
+      } catch (e) {
+        return false;
+      }
+    }
+
+    return false;
+  }
+
+  closeTab(tabId: TabId) {
+    chrome.tabs.remove(tabId);
+  }
+
+  setData(tabId: TabId, key: any, value: any) {
+    const tab = this.getTab(tabId);
+
+    if (!tab) {
+      return;
+    }
+    // null safe for firefox as in firefox get/set happen before tab updation whereas
+    // in chrome get/set happens after tab updation
+    if (tab[DataScope.TAB]) {
+      tab[DataScope.TAB][key] = value;
+    } else {
+      tab[DataScope.TAB] = { [key]: value };
+    }
+  }
+
+  getData(tabId: TabId, key: any, defaultValue?: any) {
+    const tab = this.getTab(tabId);
+
+    if (!tab) {
+      return;
+    }
+
+    return tab[DataScope.TAB]?.[key] || defaultValue;
+  }
+
+  removeData(tabId: TabId, key: any) {
+    const tab = this.getTab(tabId);
+
+    if (!tab || !tab[DataScope.TAB]) {
+      return;
+    }
+
+    delete tab[DataScope.TAB][key];
+  }
+}
+const tabService = TabService.getInstance();
```

Comment:
Each ES module is executed once, so already a singleton. No need to write `getInstance` logic.

```js
export const tabService = new TabService();
```